test/params_api_test.c

```diff
@@ -0,0 +1,391 @@
+        || !TEST_true(OSSL_PARAM_get_double(param, &d)))
+        return 0;
+
+    /* Check signed types */
+    if (bit32) {
+        copy_to_le(buf, &i32, sizeof(i32));
+        sz = sizeof(i32) < width ? sizeof(i32) : width;
+        if (!TEST_mem_eq(buf, sz, cmp, sz))
+            return 0;
+    }
+    copy_to_le(buf, &i64, sizeof(i64));
+        sz = sizeof(i64) < width ? sizeof(i64) : width;
+    if (!TEST_mem_eq(buf, sz, cmp, sz))
+        return 0;
+    if (sizet) {
+        copy_to_le(buf, &s, sizeof(s));
+        sz = sizeof(s) < width ? sizeof(s) : width;
+        if (!TEST_mem_eq(buf, sz, cmp, sz))
+            return 0;
+    }
+
+    /* Check reals, remembering that they can lose accuracy */
+    if (param->buffer_size <= 6) {
+        i64 = (int64_t)d;
+        copy_to_le(buf, &i64, sizeof(i64));
+        if (!TEST_mem_eq(buf, param->buffer_size, cmp, param->buffer_size))
+            return 0;
+    }
+
+    /* Check a widening write if possible */
+    if (sizeof(size_t) > width)
+        if (!TEST_true(OSSL_PARAM_set_int32(param, 12345))
+            || !TEST_true(OSSL_PARAM_get_size_t(param, &s))
+            || !TEST_size_t_eq(s, 12345))
+            return 0;
+    return 1;
+}
+
+/*
+ * This macro defines a basic test case that sets a param of a specified
+ * integral type, attempts to read the value back and then calls the above
+ * function to verify that the params can be read as other types.
+ *
+ * Since this macro doesn't know the exact sizes of everything, all the real
+ * work is done via byte buffers which are converted to machine order or little
+ * endian for comparisons.  Narrower values are best compared using little
+ * endian because their values and positions don't change.
+ */
+#define TEST_INT(name, type) \
+    static int test_param_##name(int n) \
+    { \
+        type in, out; \
+        unsigned char buf[MAX_LEN], le[MAX_LEN], cmp[sizeof(type)]; \
+        const size_t len = raw_values[n].len >= sizeof(type) \
+                           ? sizeof(type) : raw_values[n].len; \
+        OSSL_PARAM param = OSSL_PARAM_##name("a", NULL); \
+        \
+        memset(buf, 0, sizeof(buf)); \
+        memset(le, 0, sizeof(le)); \
+        copy_be_to_native(buf, raw_values[n].value, len); \
+        swap_copy(le, raw_values[n].value, len); \
+        memcpy(&in, buf, sizeof(in)); \
+        param.buffer = &out; \
+        if (!TEST_true(OSSL_PARAM_set_##name(&param, in))) \
+            return 0; \
+        copy_to_le(cmp, &out, sizeof(out)); \
+        if (!TEST_mem_eq(cmp, len, le, len)) \
+            return 0; \
+        in = 0; \
+        param.buffer = buf; \
+        if (!TEST_true(OSSL_PARAM_get_##name(&param, &in))) \
+            return 0; \
+        copy_to_le(cmp, &in, sizeof(in)); \
+        if (!TEST_mem_eq(cmp, sizeof(in), le, sizeof(in))) \
+            return 0; \
+        param.buffer = &out; \
+        return test_param_type_extra(&param, le, sizeof(type)); \
+    }
```

Comment:
This is far too much code to be in a macro. Debugging this in the event of a failure will be impossible.

Reply:
It was :)

I'll look at expanding the macro magicks into real code.